cdap-api/src/main/java/io/cdap/cdap/api/app/Application.java

```diff
@@ -32,4 +32,23 @@
    * @param context Used to access the environment, application configuration, and application (deployment) arguments
    */
   void configure(ApplicationConfigurer configurer, ApplicationContext<T> context);
+
+  /**
+   * Returns if application supports config upgrade or not.
+   */
+  default boolean isUpgradeAllowed() {
+    return false;
+  }
+
+  /**
+   * Upgrades given configuration.
+   *
+   * @param config Configuration, most likely older, to be upgraded.
+   * @param applicationUpdateContext Used to access methods helpful for operation like upgrade on a config such as get
+   * latest plugin versions.
+   */
+  default ApplicationUpdateResult<T> upgradeConfig(T config, ApplicationUpdateContext applicationUpdateContext)
```

Comment:
Call it `updateConfig`. Also, don't pass in `T config`. Always access it from the update context.

Reply:
Sorry, I had that change in my local client before you started reviewing. Done.